pkg/console/config/config.go

```diff
@@ -1,169 +1,136 @@
 }
 
-func (w *watch) Run(done <-chan struct{}) error {
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return fmt.Errorf("could not create fsnotify.Watcher: %w", err)
+func (c *configWatch) AddToFilewatch(watch filewatch.Watch) error {
+	if err := watch.Add(c.configDir, c.reloadConfig); err != nil {
+		return err
 	}
-	// watcher.Close() never returns an error
-	defer func() { _ = watcher.Close() }()
-
-	err = watcher.Add(w.tlsProfilePath)
-	if err != nil {
-		return fmt.Errorf("could not add watch: %w", err)
-	}
-
-	err = watcher.Add(w.certsPath)
-	if err != nil {
-		return fmt.Errorf("failed to watch Certificate: %w", err)
-	}
-
-	err = watcher.Add(w.keyPath)
-	if err != nil {
-		return fmt.Errorf("failed to watch Certificate key: %w", err)
-	}
-
-	return w.processEvents(watcher, done)
-}
-
-func (w *watch) IsRunning() bool {
-	return w.running.Load()
+	return watch.Add(c.certAndKeyDir, c.reloadCertificate)
 }
 
-func (w *watch) Reload() {
-	w.reloadTlsProfile()
-	w.reloadCertificate()
+func (c *configWatch) Reload() {
+	c.reloadConfig()
+	c.reloadCertificate()
 }
 
-func (w *watch) GetConfig() (*tls.Config, error) {
-	w.lock.RLock()
-	defer w.lock.RUnlock()
+func (c *configWatch) GetTlsConfig() (*tls.Config, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
 
-	if w.tlsProfileError != nil {
-		return nil, w.tlsProfileError
+	if c.tlsProfileError != nil {
+		return nil, c.tlsProfileError
 	}
-	if w.certError != nil {
-		return nil, w.certError
+	if c.certError != nil {
+		return nil, c.certError
 	}
 
 	return &tls.Config{
-		CipherSuites: w.ciphers,
-		MinVersion:   w.minTlsVersion,
-		Certificates: []tls.Certificate{*w.certificate},
+		CipherSuites: c.ciphers,
+		MinVersion:   c.minTlsVersion,
+		Certificates: []tls.Certificate{*c.certificate},
 	}, nil
 }
 
-func (w *watch) processEvents(watcher *fsnotify.Watcher, done <-chan struct{}) error {
-	w.running.Store(true)
-	defer w.running.Store(false)
-
-	for {
-		select {
-		case <-done:
-			return nil
-
-		case event, ok := <-watcher.Events:
-			if !ok {
-				return nil
-			}
-			w.handleEvent(event)
-
-		case err, ok := <-watcher.Errors:
-			if !ok {
-				return nil
-			}
-			if err != nil {
-				return err
-			}
-		}
+func (c *configWatch) GetUseServiceAccountTokens() (bool, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	if c.useSaTokensError != nil {
+		return false, c.useSaTokensError
 	}
+	return c.useSaTokens, nil
 }
 
-func (w *watch) handleEvent(event fsnotify.Event) {
-	const modificationEvents = fsnotify.Create | fsnotify.Write | fsnotify.Remove
-	if event.Op&modificationEvents == 0 {
-		return
-	}
+func (c *configWatch) reloadConfig() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
 
-	switch {
-	case strings.HasPrefix(w.tlsProfilePath, event.Name):
-		w.reloadTlsProfile()
-	case strings.HasPrefix(w.certsPath, event.Name):
-		w.reloadCertificate()
-	case strings.HasPrefix(w.keyPath, event.Name):
-		w.reloadCertificate()
-	}
+	c.reloadTlsProfileLocked()
```

Comment:
why does these functions have "Locked" in their names?

Reply:
Because they are called in a critical section when `c.lock` is locked.
Now that I think about it, the whole function does not need to be called with locked lock. I will optimize it in a future PR.